### PR TITLE
Change pinned timestamp file format stored in remote store

### PR DESCRIPTION
**Pull request by sachinpkale**
### Description
- As part of https://github.com/opensearch-project/OpenSearch/pull/15180, we introduced a way to pin/unpin timestamp and stored the end state in the configured remote store.
- But storing all the pinned timestamps in a file can have concurrency issues even if we create immutable files to store the info.
- While debugging, reading these pinned timestamp files in remote is also not always possible if the content is encrypted.
- To tackle these problems, in this PR, we change the way to store pinned timestamp data. Instead of storing all the timestamps in one file, we create separate, empty file per `<timestamp, pinning entity>` tuple.
- We name the file as `<pinning entity>__<timestamp>` which makes debugging easier just by listing the files.

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/remote/RemoteStoreUtils.java [hunk 1]
```diff
@@ -542,8 +542,11 @@ public static List<String> filterOutMetadataFilesBasedOnAge(
         if (RemoteStoreSettings.isPinnedTimestampsEnabled() == false) {
             return new ArrayList<>(metadataFiles);
         }
-        long maximumAllowedTimestamp = lastSuccessfulFetchOfPinnedTimestamps - RemoteStoreSettings.getPinnedTimestampsLookbackInterval()
-            .getMillis();
+        // We allow now() - loopback interval to be pinned. Also, the actual pinning can take at most loopback interval
+        // This means the pinned timestamp can be available for read after at most (2 * loopback interval)
+        long maximumAllowedTimestamp = lastSuccessfulFetchOfPinnedTimestamps - (2 * RemoteStoreSettings
+            .getPinnedTimestampsLookbackInterval()
+            .getMillis());
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/remote/RemoteStoreUtils.java:549]: Have we factored in GC pauses?

Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/remote/RemoteStoreUtils.java:549]: Yes, we fail pinning the timestamp if it takes more than lookup interval, so GC pauses, remote store latency etc. are factored

File: server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java [hunk 2]
```diff
@@ -129,101 +109,106 @@ public void pinTimestamp(long timestamp, String pinningEntity, ActionListener<Vo
                 "Timestamp to be pinned is less than current timestamp - value of cluster.remote_store.pinned_timestamps.lookback_interval"
             );
         }
-        updatePinning(pinnedTimestamps -> pinnedTimestamps.pin(timestamp, pinningEntity), listener);
+        long startTime = System.nanoTime();
+        try {
+            logger.debug("Pinning timestamp = {} against entity = {}", timestamp, pinningEntity);
+            blobContainer.writeBlob(getBlobName(timestamp, pinningEntity), new ByteArrayInputStream(new byte[0]), 0, true);
```
[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java:115]: Ideally, the `writeBlob` method should handle closing the stream. For non empty files as well, we provide an opened InputStream to writeBlob.

File: server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java [hunk 3]
```diff
@@ -129,101 +109,106 @@ public void pinTimestamp(long timestamp, String pinningEntity, ActionListener<Vo
...
+            long elapsedTime = System.nanoTime() - startTime;
+            if (elapsedTime > RemoteStoreSettings.getPinnedTimestampsLookbackInterval().nanos()) {
+                String errorMessage = String.format(
+                    Locale.ROOT,
+                    "Timestamp pinning took %s nanoseconds which is more than limit of %s nanoseconds, failing the operation",
+                    elapsedTime,
+                    RemoteStoreSettings.getPinnedTimestampsLookbackInterval().nanos()
+                );
+                unpinTimestamp(timestamp, pinningEntity, ActionListener.wrap(() -> listener.onFailure(new RuntimeException(errorMessage))));
```
Comment (linuxpi) [inline server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java:124]: - can it happen that pin passes and unpin fails? do we need to plan about some background cleanup thread in future?
- should the unpinning we done on an async executor thread, so as to not block the caller for long? if pinning took longer i guess there is a high change unpinning also will take longer

File: server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java [hunk 4]
```diff
@@ -129,101 +109,106 @@ public void pinTimestamp(long timestamp, String pinningEntity, ActionListener<Vo
...
         try {
-            return pinnedTimestampsBlobStore().read(remotePinnedTimestamps);
+            logger.debug("Unpinning timestamp = {} against entity = {}", timestamp, pinningEntity);
+            String blobName = getBlobName(timestamp, pinningEntity);
+            if (blobContainer.blobExists(blobName)) {
+                blobContainer.deleteBlobsIgnoringIfNotExists(List.of(blobName));
+                listener.onResponse(null);
+            } else {
+                String errorMessage = String.format(Locale.ROOT, "Timestamp: %s is not pinned by entity: %s", timestamp, pinningEntity);
+                listener.onFailure(new IllegalArgumentException(errorMessage));
```
Comment (linuxpi) [inline server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java:208]: the listener might not log the error, should we log it?

File: server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java (hunk 3 above)
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java:124]: >can it happen that pin passes and unpin fails? do we need to plan about some background cleanup thread in future?

**Review by linuxpi — APPROVED** [review]

**Comment by sachinpkale** [comment]
Changes are mostly covered in the integ tests. CodeCov is failing as it does not consider coverage from integ tests.

**Pull request merged**